Improved performance of newHashMap etc.

spoenemann opened the pull request:
* When `newHashMap` or `newLinkedHashMap` is used without parameters, I would expect the same result that I get with `new HashMap` or `new LinkedHashMap`: maps initialized with the default capacity (16). Previously they were initialized with a capacity of 1, which means that rehashing was necessary even if you put just a few elements into them.

 * The `LinkedHashMap` constructor is now called with the capacity value, which is 4/3 &times; the initial size due to the load factor. Previously the initial size was used, which means that rehashing was necessary for the initial content.

szarnekow commented:
We should probably add a no-args overload that avoid the construction of the empty array?

spoenemann commented:
@szarnekow I would handle that in a separate PR.

I added size checks to `newArrayList` and `newHashSet` and switched to using the constructors directly instead of calling the Guava functions that call the constructors.

Interestingly, size checks were already present in `newLinkedList`, `newLinkedHashSet`, and `newTreeSet`, but not for `newArrayList`, `newHashSet`, `newHashMap`, and `newLinkedHashMap`.

Note that for `newArrayList`, which is probably the most popular method here, the initial capacity for a zero-length argument now changes from `5` (the capacity for an array of zero elements) to `10` (the default capacity of `new ArrayList<>()`). This might have quite some impact on applications that use Xbase, especially for Xtext. Any opinions on this?

svenefftinge commented:
Having the overloads in this PR makes sense since you can always just delegate.

svenefftinge commented:
Otherwise, +1 for going with the JDK defaults for `newArrayList`.

spoenemann commented:
@szarnekow @svenefftinge I added no-arg overloads.

szarnekow commented on org.eclipse.xtext.xbase.lib/src/org/eclipse/xtext/xbase/lib/CollectionLiterals.java:
```diff
@@ -231,9 +311,34 @@
+	 * @return a new {@link TreeMap}
+	 */
+	@Pure
+	public static <K, V> TreeMap<K, V> newTreeMap(Comparator<? super K> comparator) {
```
Should we add an overload newTreeMap() where we define K as K extends Comparable<? super K>

spoenemann replied:
Sure, but I'd like to merge this PR before my vacation ;-)
Your proposal is about new API, while this PR is about the performance of existing API.

szarnekow replied:
👍

cdietrich commented:
this breaks gazillions of content assist tests. please fix

svenefftinge commented:
Also, it causes problems for everyone using >2.13 in workspace and a version before this change in IDE. Maybe it is not worth and we should just roll back the change (at least the overloads).

cdietrich commented:
No needed . Fixed the tests . Eclipse issue is fixed by update . As stated https://www.eclipse.org/forums/index.php/t/1088834/ an update will fix

spoenemann commented:
> Also, it causes problems for everyone using >2.13 in workspace and a version before this change in IDE.

This is the same situation as with #21. It's a tradeoff between performance optimization and bytecode compatibility.
